### PR TITLE
feat: converted code to typescript

### PR DESCRIPTION
**Pull request by DiogoRDuarte**
## Because
- The current implementation does not make usage of Typescript, something that can hurt the development experience.
- These changes were previously suggested in another PR (now closed). Given its magnitude, it was decided to break it into smaller, more manageable, ones. Kudos to @JoaoTMDias ✌️

### Refactor
- Converts the codebase to typescript (except the JSX components which were left as plain JSX without any type inferrence).

### CODE REVIEW

**Review by JoaoTMDias — CHANGES_REQUESTED** [review]
Overall looks ok, what I feel it is needed is better jsdoc docs for all functions in the project. For open-source projects this is essential to read and understand code.

There's a vscode extension called "Document This" that helps with most of the properties necessary with jsdoc. But functions need short and concise descriptions.

**Comment by DiogoRDuarte** [comment]
@JoaoTMDias Documentation added!

**Review by JoaoTMDias — APPROVED** [review]
Well done 👏

File: src/AutoVizuA11y.tsx [hunk 1]
```diff
@@ -21,6 +18,71 @@ import { descriptionsChanger } from "./components/descriptions/Descriptions";
...
+	insights: string;
+	descriptor?: string;
+	multiSeries?: string;
+	autoDescriptions?: AutoDescriptionsProps;
+	manualDescriptions?: ManualDescriptionsProps;
+	children: React.ReactNode;
+};
+
+/**
+ * AutoVizuA11y component, that adds screen-reader accessibility to wrapped charts.
```
Comment (ritacosta0) [inline src/AutoVizuA11y.tsx:48]: Please remove the comma.

File: src/AutoVizuA11y.tsx [hunk 2]
```diff
@@ -21,6 +18,71 @@ import { descriptionsChanger } from "./components/descriptions/Descriptions";
...
+ *
+ * @param {AutoVizuA11yProps} {
+ * 	type, - type of chart;
+ * 	descriptor, - descriptor of each data element;
+ * 	selectorType, - HTML type or  classname of the data elements in the DOM;
```
Comment (ritacosta0) [inline src/AutoVizuA11y.tsx:53]: Extra space between "or" and "classname"

File: src/ShortcutGuide.tsx [hunk 3]
```diff
@@ -8,7 +8,12 @@
 import React, { useRef, useEffect } from "react";
 import "./style/ShortcutGuideStyle.css";
 
-const ShortcutGuide = () => {
+/**
+ * Component that renders the list (also visual) of all AutoVizuA11y shortcuts.
```
Comment (ritacosta0) [inline src/ShortcutGuide.tsx:12]: Does the "also visual" matter?

File: src/components/descriptions/Descriptions.ts [hunk 4]
```diff
@@ -5,11 +5,37 @@
...
+type AutoDescriptionsProps = {
+	dynamicDescriptions?: boolean;
+	apiKey: string;
+	model?: string;
+	temperature?: number;
+};
 
-	//When pressed reads the smaller description
+/**
+ * Handles the longer and shorter description change when Alt+B or Alt+S are pressed, respectively.
```
Comment (ritacosta0) [inline src/components/descriptions/Descriptions.ts:16]: The alt+B and alt+S command is not for MacOS. Can you add the "option" detail here and everywhere you mention shortcuts in the documentation?

File: src/components/navigation/LevelNavigation.ts [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
+ * You should have received a copy of the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
+ * Other licensing options may be available, please reach out to data-viz@feedzai.com for more information.
+ */
+
+import { wiper } from "../../utils/wiper";
+import { addDataNavigation } from "./AddNavigation";
+
+/**
+ * Changes navigation based on key press.
```
Comment (ritacosta0) [inline src/components/navigation/LevelNavigation.ts:12]: What do we mean by "changes navigation" here?

File: src/components/navigation/LevelNavigation.ts [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+
+	return;
+}
+
+interface ExtendedHTMLElement extends HTMLElement {
+	pastFocus?: HTMLElement | null;
+}
+
+/**
+ * Shows the ShortcutGuide and gives it keyboard focus.
```
Comment (ritacosta0) [inline src/components/navigation/LevelNavigation.ts:155]: I suggest using the word "display" here

File: src/components/navigation/LevelNavigation.ts [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+	for (let i = 0; i < allCharts.length; i++) {
+		allCharts[i].removeAttribute("tabIndex");
+	}
+	wiper(ref);
+	// const defaultSelectorType = { element: "defaultElement", className: "defaultClassName" };
+	addDataNavigation(ref, selectorType, selectedSeries);
+}
+
+/**
+ * Enables navigation on another data series.
```
Comment (ritacosta0) [inline src/components/navigation/LevelNavigation.ts:252]: It's more something like "Enables navigation between data series", no?

File: src/components/shortcuts/OverallComparer.ts [hunk 8]
```diff
@@ -5,21 +5,36 @@
...
 
-import { median } from "../../utils/maths";
-import { getOrdinalNumber } from "../../utils/maths";
+import { median, getOrdinalNumber } from "../../utils/maths";
 
-//makes the comparison between the focusedData and the overall data
-export function overallComparer(event, alertDiv, insights, arrayConverted, focusedData) {
-	const { nativeEvent } = event;
+/**
+ * Produces a message based on corner cases and calls comparer otherwise.
```
Comment (ritacosta0) [inline src/components/shortcuts/OverallComparer.ts:11]: What exactly do you mean about the "message based on corner cases"? Is that the focus of the function or is this secondary?

File: src/components/shortcuts/Skip.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
+/**
+ * This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, version 3 of the License.
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
+ * You should have received a copy of the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
+ * Other licensing options may be available, please reach out to data-viz@feedzai.com for more information.
+ */
+
+/**
+ * Enables one to jump to the end/beginning of a chart.
```
Comment (ritacosta0) [inline src/components/shortcuts/Skip.ts:9]: Just for the phrasing to be more aligned with the others, I suggest writing something like "Enables the navigation to the end/beginning of a chart"

File: src/utils/maths.ts [hunk 10]
```diff
@@ -5,13 +5,25 @@
...
+ * @return {number} Rounded number.
+ */
+export function rounding(number: number) {
 	return Math.round(number * 100) / 100;
 }
 
-//calculates the median value
-export function median(arr) {
+/**
+ * Calculates the median value of a number array.
```
Comment (ritacosta0) [inline src/utils/maths.ts:20]: a number array -> an array of numbers

File: src/utils/maths.ts [hunk 11]
```diff
@@ -21,8 +33,14 @@ export function median(arr) {
 	}
 }
 
-//returns the ordinal number based on a number
-export function getOrdinalNumber(number) {
+/**
+ * Determines the ordinal version of a number.
```
Comment (ritacosta0) [inline src/utils/maths.ts:37]: Maybe something like "Transforms number to ordinal". Wdyt?

File: src/ShortcutGuide.tsx (hunk 3 above)
Comment (DiogoRDuarte) [reply src/ShortcutGuide.tsx:12]: I agree it can be cut

**Pull request merged**